docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+to control versions of `TrainingRuntime` and enable rolling updates.
+
+We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
+exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
+If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
+otherwise it will use the `ClusterTrainingRuntime`.
+
+```golang
+type TrainingRuntime struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Framework specific parameters.
+    MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+    // Number of nodes to execute training.
+    NumNodes int `json:"numNodes,omitempty"`
+
+    // JobSet spec.
+    JobSetSpec *batchv1.JobSetSpec `json:",inline"`
+
+    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
+    GangScheduler *GangScheduler `json:"gangScheduler,omitempty"`
+}
+
+// One of the specs can be selected.
+type MLSpec struct {
+
+    // Custom Spec for Torch
+    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+    // Custom Spec for MPI
+    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+```
+
+### The Gang Scheduler API
+
+Gang scheduler plugin is used to create the appropriate `PodGroup` for Volcano or scheduler plugins.
+
+```golang
+type GangScheduler struct {
+    // Plugin for gang scheduling.
+    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`
+
+    // Time threshold to schedule PodGroup for gang scheduling.
+    ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`
+}
+
+type GangSchedulerPlugin string
+
+const (
+    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+    GangSchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"
+)
+```
+
+### The Torch Spec API
+
+The `TorchSpec` API represents the configuration for the PyTorch distributed training. This configuration
+allows platform engineers to explicitly configure `torchrun` setting.
+
+The distributed parameters are taken from the
+[PyTorch distributed launch run](https://github.com/pytorch/pytorch/blob/94dc3253a0fefbfb95fbe467ddd628e4c2eb08d7/torch/distributed/run.py).
+
+For Elastic Training we will always pass the following parameters:
+
+```bash
+--rdzv-backend=c10d
+
+--rdzv-id will be set automatically.
+
+--rdzv-endpoint will always point to the node-0 Pod.
+```
+
+Since the [etcd and etcd-v2 are legacy rendezvous](https://pytorch.org/docs/stable/elastic/run.html#note-on-rendezvous-backend),
+we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.
+
+```golang
+// TorchSpec represents the configuration for PyTorch.
+type TorchSpec struct {
+
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // Used for single-node multi-worker training
+    Standalone bool `json:"standalone,omitempty"`
```

Comment:
```suggestion
    Standalone bool `json:"standalone"`
```

Let's handle this bool field as non omitempty field so that we can set false as a default.

Reply:
@tenzen-y In that case do we need to configure: `standalone: False` in every Torch runtime ?

Reply:
The answer is no. Users can skip to specify the `standalone: false` when they do not want to use the standalone mode.

Reply:
Since this can be omitted in the TrainingRuntime YAML, why we don't want to set it as:
```go
 Standalone bool `json:"standalone,omitempty"`
 ```